File: 77. Combinations.py
# timeout

class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        self.res = []
        self.DFS([], 1, n, k)
        return self.res
        
    def DFS(self, path, start, n, k):
        if len(path) == k:
            self.res += [path]
            return
        for i in xrange(start, n+1):
            if not path or (path and i > path[len(path)-1]):
                self.DFS(path + [i], start, n, k)

File: 77. Combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        DFS(res, {}, 1, n, k);
        return res;
    }
    
    void DFS(vector<vector<int>>& res, vector <int> path, int start, int n, int k)
    {
        if (path.size() == k)
        {
            res.push_back(path);
            return;
        }
        for (int i = start; i <= n; i++)
        {
            path.push_back(i);
            DFS(res, path, i+1, n, k);
            path.pop_back();
        }
    }
};
